File: Implementierung/Backend/src/main/java/edu/kit/informatik/model/userdata/courses/SeatArrangement.java
package edu.kit.informatik.model.userdata.courses;

import edu.kit.informatik.model.DataObject;
import edu.kit.informatik.model.User;
import edu.kit.informatik.model.userdata.interactions.Participant;
import edu.kit.informatik.model.userdata.rooms.Chair;
import edu.kit.informatik.model.userdata.rooms.Room;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.MapKeyJoinColumn;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.sql.Timestamp;
import java.util.Map;

/**
 * Entität einer Sitzordnung
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Getter
@NoArgsConstructor
@Entity
@Table(name = "seatarrangements")
public class SeatArrangement extends DataObject {

    private String name;

    @OneToOne
    private Room room;

    @ManyToOne
    private Course course;

    @ManyToMany(cascade = CascadeType.ALL)
    @JoinTable(name = "chair_participant_mapping",
            joinColumns = @JoinColumn(name = "seatarrangementId"),
            inverseJoinColumns  = @JoinColumn(name = "participantId"))
    @MapKeyJoinColumn(name = "chairId")

    private Map<Chair, Participant> seatMap;

    /**
     * Konstruktor zum Erstellen eines Objektes
     * @param user {@link User}
     * @param name Name
     * @param room {@link Room}
     * @param course {@link Course}
     * @param seatMap {@link Map} mit Key {@link Chair} und Value {@link Participant}
     * @param createdAt {@link Timestamp} der Erstellung
     * @param updatedAt {@link Timestamp} der letzten Änderung
     */
    public SeatArrangement(User user, String name, Room room, Course course, Map<Chair,
                            Participant> seatMap, Timestamp createdAt, Timestamp updatedAt) {
        super(user, createdAt, updatedAt);
        this.name = name;
        this.room = room;
        this.course = course;
        this.seatMap = seatMap;
    }

    /**
     * Setzen der Sitzplatz-Map
     * @param seatMap {@link Map} mit Key {@link Chair} und Value {@link Participant}
     */
    public void setSeatMap(Map<Chair, Participant> seatMap) {
        this.seatMap = seatMap;
    }

    /**
     * Setzen des Namens
     * @param name Name
     */
    public void setName(String name) {
        this.name = name;
    }
}

File: Implementierung/frontend/paschi/src/model/userdata/courses/SeatArrangement.ts
import {User} from "@/model/User";
import {RoomObject} from "@/model/userdata/rooms/RoomObject";
import {Participant} from "@/model/userdata/interactions/Participant";
import {Course} from "@/model/userdata/courses/Course";
import {Room} from "@/model/userdata/rooms/Room";
import {DataObject} from "@/model/DataObject";

export class SeatArrangement extends DataObject {

  private readonly _user: User
  private _name: string;
  private _seatMap: Map<RoomObject, Participant>;
  private readonly _course: Course;
  private readonly _room: Room;

  constructor(id: string | undefined, localId: number, user: User, name: string, course: Course, room: Room) {
    super(id, localId);
    this._user = user;
    this._name = name;
    this._seatMap = new Map<RoomObject, Participant>();
    this._course = course;
    this._room = room;
  }

  isVisible(): boolean {
    return this.room.visible;
  }

  getParticipantForSeat(seat: RoomObject): Participant | undefined {
    return this._seatMap.get(seat);
  }

  setSeat(seat: RoomObject, participant: Participant) {
    this._seatMap.set(seat, participant);
    this.update();
  }

  removeSeat(seat: RoomObject) {
    this._seatMap.delete(seat);
    this.update();
  }

  getAllStudents(): Participant[] {
    return this.course.participants;
  }

  getStudentsNotAssigned(): Participant[] {
    return this.getAllStudents().filter(student => {
      for (const value of this._seatMap) {
        if (value[1].getId === student.getId) {
          return false;
        }
      }
      return true;
    });
  }

  get user(): User {
    return this._user;
  }

  get name(): string {
    return this._name;
  }

  get seatMap(): Map<RoomObject, Participant> {
    return this._seatMap;
  }

  get course(): Course {
    return this._course;
  }

  get room(): Room {
    return this._room;
  }

  set name(value: string) {
    this._name = value;
    this.update();
  }

  set seatMap(value: Map<RoomObject, Participant>) {
    this._seatMap = value;
  }
}
